src/api.ts

```diff
@@ -195,69 +236,264 @@ function createApi({
+            'JOIN Tile ON TileData.tileHash = Tile.tileHash ' +
+            'JOIN Tileset ON Tile.tilesetId = Tileset.id ' +
+            'WHERE Tileset.id = :tilesetId AND Tile.quadKey = :quadKey'
+        )
+        .get({ tilesetId, quadKey })
+
+      async function fetchOnlineResource(): Promise<
+        | {
+            data: Buffer
+            etag?: string
+          }
+        | undefined
+      > {
+        const upstreamTileUrl = getUpstreamTileUrl({
+          tilesetId,
+          zoom,
+          x,
+          y,
+        })
+
+        // TODO: Need to check if we can make online requests too
+        if (upstreamTileUrl) {
+          const response = await upstreamRequestsManager.getUpstream<Buffer>({
+            url: upstreamTileUrl,
+            etag: tile?.etag,
+            responseType: 'buffer',
+          })
+
+          if (response) {
+            await api.putTile({
+              tilesetId,
+              zoom,
+              x,
+              y,
+              data: response.data,
+              etag: response.etag,
+            })
+
+            return { data: response.data, etag: response.etag }
+          }
+        }
+      }
+
+      if (tile) {
+        fetchOnlineResource().catch(() => {})
+      } else {
+        const response = await fetchOnlineResource()
+
+        // TODO: Should this throw if false?
+        if (response) {
+          tile = response
+        }
+      }
+
+      if (!tile) {
+        // TODO: Improve error handling here?
+        throw new NotFoundError(
+          `Tileset id = ${tilesetId}, [${zoom}, ${x}, ${y}]`
+        )
+      }
+
+      return {
+        data: tile.data,
+        // TODO: This never returns a Last-Modified header but seems like the endpoint would want it to if possible?
+        // Would require changing the return type of UpstreamRequestsManager.getUpstream
+        headers: { ...headers(tile.data), Etag: tile.etag },
       }
-      return tilestore.getTile(zoom, x, y)
+    },
+
+    async putTile({ tilesetId, zoom, x, y, data, etag }) {
+      const upstreamTileUrl = getUpstreamTileUrl({
+        tilesetId,
+        zoom,
+        x,
+        y,
+      })
+
+      const quadKey = tileToQuadkey([x, y, zoom])
+
+      const transaction = db.transaction(() => {
+        const tileHash = hash(data).toString('hex')
+
+        db.prepare<{
+          tileHash: string
+          tilesetId: string
+          data: Buffer
+        }>(
+          'INSERT INTO TileData (tileHash, tilesetId, data) VALUES (:tileHash, :tilesetId, :data)'
+        ).run({ tileHash, tilesetId, data })
+
+        db.prepare<{
+          etag?: string
+          tilesetId: string
+          upstreamUrl?: string
+        }>(
+          'UPDATE Tileset SET (etag, upstreamUrl) = (:etag, :upstreamUrl) WHERE id = :tilesetId'
+        ).run({
+          etag,
+          tilesetId,
+          upstreamUrl: upstreamTileUrl,
+        })
```

Comment:
can you elaborate? i thought we needed to persist of the etag+upstream url that we use to make the most recent upstream request.

also, think i'm getting a little mixed up with having both `upstreamTileUrls` and `upstreamUrl` as fields in the `Tileset` table 😅

Reply:
Yeah, I hadn't spotted this before. We need an etag for each tile - that way we know whether to download the new one or not. The etag on the tileset table is the etag for the last time the _tileset_ (e.g. the tilejson) was fetched. We could just store the upstream URL of each tile as well (like we do for all other resources) but because the URL for each tile can be derived from the tile URL template, as a performance / storage optimization we just store the template in the tileset table (as upstreamTileUrls) so we avoid the need to store the URL for each tile in the Tile table.

Reply:
ah i see. so we need to update the schema in this case to add `etag` as a field for `TileData` right?

Reply:
yeah, although I think for `Tile` table rather than `TileData` though.